### ISSUE TITLE
_id from insert

### ISSUE THREAD

**Issue by tcurdt** [marcello3d/node-mongolian#38]
shouldn't an insert return the _id?

**Comment by tcurdt** [comment]
Looking at the tests I tried this

```
  var a = {
    items: [
    { id: '1', title: 'Item1' },
    { id: '2', title: 'Item2' } ] };

  orders.insert(a, function(b) {
    console.dir(b);
```

but for some reason `b` is null.

Also wondering about the error handling on inserts/updates. Does `null` mean error?

**Comment by tcurdt** [comment]
Figured it out. The callback signature is actually `function(err, doc)` ...but doesn't that mean the tests are out of date?

**Comment by marcello3d** [comment]
The _id is generated and added to the object by the driver, not the server, so immediately after calling foo.insert() or foo.save(), you should be able to retrieve _id from the object. Passing a callback to insert() will actually wait until the server confirms that the object has been inserted (though you're not required to wait for this).

The tests are using vowsjs, which automatically handles the error parameter, hence the error parameter not showing up. I'm not super happy with vowsjs (it's a little confusing/hard to follow), so I may change this in the future.

Thanks for the notes.

**Comment by tcurdt** [comment]
Ah thanks for the explanation. Will give it a try without the callback and see how that works out.

**Comment by tcurdt** [comment]
Hmmm. So you are saying this should work?

```
  var obj = { ... }
  collection.insert(obj);

  if (obj._id) {
    // all good
  } else {
    // nope
  }
```

...because it doesn't seem to.

**Comment by marcello3d** [comment]
collection.insert() will _always_ generate an _id if one isn't set already.
It's not an indication of insertion success, but it can be used to further
modify the object. To verify the insert was successful, you do need to use a
callback (since it's an asynchronous network-based operation).

**Comment by tcurdt** [comment]
This is not about whether the insert was successful or not - I just want the _id ...and somehow I only get it when I use the callback. From your explanation I would think the above snippet should work too.

**Comment by marcello3d** [comment]
It should work... I'm not really sure what to say.

Here is the code for insert, it's not particularly complicated:
https://github.com/marcello3d/node-mongolian/blob/master/lib/collection.js#L59

Are you getting an exception?

**Comment by tcurdt** [comment]
Now that is indeed odd. Yeah, from the code it should do exactly that.
Will debug and get back to you.

cheers,
Torsten

**Comment by tcurdt** [comment]
I've tested it outside of my app and there it works.

```
var Mongolian = require("mongolian")

module.exports = {
  "rerun test":function(test) {
    var mongo = new Mongolian(),
        db = mongo.db('test'),
        order = db.collection('order');

    console.log('new mongolian');
    function foo() {
      var o = { something: 'foo' }
      order.insert(o);
      console.dir(o);
    }

    foo()
  }
}
```

So it must be something else. Sorry for the noise.
